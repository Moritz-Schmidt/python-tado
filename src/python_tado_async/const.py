"""Constants for the asynchronous Python API for Tado."""

from enum import Enum

CONST_MODE_OFF = "OFF"
CONST_MODE_SMART_SCHEDULE = "SMART_SCHEDULE"
CONST_MODE_AUTO = "AUTO"
CONST_MODE_COOL = "COOL"
CONST_MODE_HEAT = "HEAT"
CONST_MODE_DRY = "DRY"
CONST_MODE_FAN = "FAN"


class HttpMethod(Enum):
    """HTTP methods."""

    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    DELETE = "DELETE"
