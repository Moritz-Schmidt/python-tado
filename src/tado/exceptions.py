"""Exceptions for Tado API."""


class TadoException(Exception):
    """Base exception for Tado API."""


class TadoConnectionError(TadoException):
    """Tado connection exception."""


class TadoAuthenticationError(TadoException):
    """Tado authentication exception."""


class TadoServerError(TadoException):
    """Tado server exception."""


class TadoBadRequestError(TadoException):
    """Tado bad request exception."""


class TadoForbiddenError(TadoException):
    """Tado forbidden exception."""
