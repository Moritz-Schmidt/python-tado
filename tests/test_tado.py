"""Tests for the Python Tado."""

import asyncio
import time
from typing import Any
from unittest.mock import AsyncMock, MagicMock, patch

import aiohttp
import pytest
from aiohttp import ClientResponse, ClientResponseError, RequestInfo
from aioresponses import CallbackResult, aioresponses
from tadoasync import (
    Tado,
)
from tadoasync.exceptions import (
    TadoBadRequestError,
    TadoConnectionError,
    TadoError,
)

from syrupy import SnapshotAssertion
from tests import load_fixture

from .const import TADO_API_URL, TADO_TOKEN_URL


async def test_create_session(
    responses: aioresponses,
) -> None:
    """Test putting in own session."""
    responses.get(
        f"{TADO_API_URL}/me",
        status=200,
        body=load_fixture("me.json"),
    )
    async with aiohttp.ClientSession() as session:
        tado = Tado(username="username", password="password", session=session)
        await tado.get_me()
        assert tado.session is not None
        assert not tado.session.closed
        await tado.close()
        assert tado.session.closed


async def test_login_success(responses: aioresponses) -> None:
    """Test login success."""
    responses.get(
        f"{TADO_API_URL}/me",
        status=200,
        body=load_fixture("me.json"),
    )
    async with aiohttp.ClientSession() as session:
        tado = Tado(username="username", password="password", session=session)
        await tado._login()
        assert tado._access_token == "test_access_token"
        assert tado._token_expiry is not None
        assert tado._token_expiry > time.time()
        assert tado._refesh_token == "test_refresh_token"


async def test_login_timeout(python_tado: Tado, responses: aioresponses) -> None:
    """Test login timeout."""
    responses.post(
        TADO_TOKEN_URL,
        exception=asyncio.TimeoutError(),
    )
    with pytest.raises(TadoConnectionError):
        await python_tado._login()


async def test_login_invalid_content_type(
    python_tado: Tado, responses: aioresponses
) -> None:
    """Test login invalid content type."""
    responses.post(
        TADO_TOKEN_URL,
        status=200,
        headers={"content-type": "text/plain"},
        body="Unexpected response",
    )

    with pytest.raises(TadoError):
        await python_tado._login()


async def test_login_client_response_error(python_tado: Tado) -> None:
    """Test login client response error."""
    mock_request_info = MagicMock(spec=RequestInfo)
    mock_response = MagicMock(spec=ClientResponse)
    mock_response.raise_for_status.side_effect = ClientResponseError(
        mock_request_info, (mock_response,), status=400
    )
    mock_response.status = 400
    mock_response.text = AsyncMock(return_value="Error message")

    async def mock_post(*args: Any, **kwargs: Any) -> ClientResponse:  # noqa: ARG001 # pylint: disable=unused-argument
        return mock_response

    with patch("aiohttp.ClientSession.post", new=mock_post), pytest.raises(
        TadoBadRequestError
    ):
        await python_tado._login()


async def test_refresh_auth_success(responses: aioresponses) -> None:
    """Test successful refresh of auth token."""
    responses.post(
        TADO_TOKEN_URL,
        status=200,
        payload={
            "access_token": "new_test_access_token",
            "expires_in": "3600",
            "refresh_token": "new_test_refresh_token",
        },
        headers={"content-type": "application/json"},
    )
    async with aiohttp.ClientSession() as session:
        tado = Tado(username="username", password="password", session=session)
        tado._access_token = "old_test_access_token"
        tado._token_expiry = time.time() - 10  # make sure the token is expired
        tado._refesh_token = "old_test_refresh_token"
        await tado._refresh_auth()
        assert tado._access_token == "test_access_token"
        assert tado._token_expiry > time.time()
        assert tado._refesh_token == "test_refresh_token"


async def test_refresh_auth_timeout(python_tado: Tado, responses: aioresponses) -> None:
    """Test timeout during refresh of auth token."""
    responses.post(
        TADO_TOKEN_URL,
        exception=asyncio.TimeoutError(),
    )
    async with aiohttp.ClientSession():
        python_tado._access_token = "old_test_access_token"
        python_tado._token_expiry = time.time() - 10  # make sure the token is expired
        python_tado._refesh_token = "old_test_refresh_token"
        with pytest.raises(TadoConnectionError):
            await python_tado._refresh_auth()


async def test_refresh_auth_client_response_error(python_tado: Tado) -> None:
    """Test client response error during refresh of auth token."""
    mock_request_info = MagicMock(spec=RequestInfo)
    mock_response = MagicMock(spec=ClientResponse)
    mock_response.raise_for_status.side_effect = ClientResponseError(
        mock_request_info, (mock_response,), status=400
    )
    mock_response.status = 400
    mock_response.text = AsyncMock(return_value="Error message")

    async def mock_post(*args: Any, **kwargs: Any) -> ClientResponse:  # noqa: ARG001 # pylint: disable=unused-argument
        return mock_response

    with patch("aiohttp.ClientSession.post", new=mock_post):
        python_tado._access_token = "old_test_access_token"
        python_tado._token_expiry = time.time() - 10  # make sure the token is expired
        python_tado._refesh_token = "old_test_refresh_token"
        with pytest.raises(TadoBadRequestError):
            await python_tado._refresh_auth()


async def test_get_me(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get me."""
    responses.get(
        f"{TADO_API_URL}/me",
        status=200,
        body=load_fixture("me.json"),
    )
    assert await python_tado.get_me() == snapshot


async def test_get_devices(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get devices."""
    responses.get(
        f"{TADO_API_URL}/homes/1/devices",
        status=200,
        body=load_fixture("devices.json"),
    )
    assert await python_tado.get_devices() == snapshot


async def test_get_mobile_devices(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get mobile devices."""
    responses.get(
        f"{TADO_API_URL}/homes/1/mobileDevices",
        status=200,
        body=load_fixture("mobile_devices.json"),
    )
    assert await python_tado.get_mobile_devices() == snapshot


async def test_get_zones(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get zones."""
    responses.get(
        f"{TADO_API_URL}/homes/1/zones",
        status=200,
        body=load_fixture("zones.json"),
    )
    assert await python_tado.get_zones() == snapshot


async def test_get_zone_states(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get zone states."""
    responses.get(
        f"{TADO_API_URL}/homes/1/zoneStates",
        status=200,
        body=load_fixture("zone_states.json"),
    )
    assert await python_tado.get_zone_states() == snapshot


async def test_get_weather(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get weather."""
    responses.get(
        f"{TADO_API_URL}/homes/1/weather",
        status=200,
        body=load_fixture("weather.json"),
    )
    assert await python_tado.get_weather() == snapshot


async def test_get_home_state(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get home state."""
    responses.get(
        f"{TADO_API_URL}/homes/1/state",
        status=200,
        body=load_fixture("home_state.json"),
    )
    assert await python_tado.get_home_state() == snapshot


async def test_get_capabilities(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get capabilities."""
    responses.get(
        f"{TADO_API_URL}/homes/1/zones/1/capabilities",
        status=200,
        body=load_fixture("capabilities.json"),
    )
    assert await python_tado.get_capabilities(1) == snapshot


async def test_reset_zone_overlay_success(
    python_tado: Tado, responses: aioresponses
) -> None:
    """Test successful reset of zone overlay."""
    zone_id = 1
    responses.delete(
        f"{TADO_API_URL}/homes/{python_tado._home_id}/zones/{zone_id}/overlay",
        status=204,
    )
    await python_tado.reset_zone_overlay(zone_id)


async def test_set_presence_success(python_tado: Tado, responses: aioresponses) -> None:
    """Test successful setting of presence."""
    presence = "HOME"
    responses.put(
        f"{TADO_API_URL}/homes/{python_tado._home_id}/presenceLock",
        status=204,
        payload={"homePresence": presence},
    )
    await python_tado.set_presence(presence)


async def test_get_device_info(
    python_tado: Tado, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test get device info."""
    responses.get(
        f"{TADO_API_URL}/devices/1/temperatureOffset",
        status=200,
        body=load_fixture("device_info.json"),
    )
    assert await python_tado.get_device_info("1", "temperatureOffset") == snapshot


async def test_set_zone_overlay_success(
    python_tado: Tado, responses: aioresponses
) -> None:
    """Test successful setting of zone overlay."""
    zone = 1
    overlay_mode = "MANUAL"
    set_temp = 20.5
    duration = 3600
    device_type = "HEATING"
    power = "ON"
    mode = "COOL"
    fan_speed = "HIGH"
    swing = "ON"

    responses.put(
        f"{TADO_API_URL}/homes/{python_tado._home_id}/zones/{zone}/overlay",
        status=200,
    )

    await python_tado.set_zone_overlay(
        zone,
        overlay_mode,
        set_temp,
        duration,
        device_type,
        power,
        mode,
        fan_speed,
        swing,
    )


async def test_request_client_response_error(python_tado: Tado) -> None:
    """Test client response error during request."""
    mock_request_info = MagicMock(spec=RequestInfo)
    mock_response = MagicMock(spec=ClientResponse)
    mock_response.raise_for_status.side_effect = ClientResponseError(
        mock_request_info, (mock_response,), status=400
    )
    mock_response.status = 400
    mock_response.text = AsyncMock(return_value="Error message")

    async def mock_get(*args: Any, **kwargs: Any) -> ClientResponse:  # noqa: ARG001 # pylint: disable=unused-argument
        return mock_response

    with patch("aiohttp.ClientSession.request", new=mock_get), pytest.raises(
        TadoBadRequestError
    ):
        await python_tado._request("me")


async def test_get_me_timeout(python_tado: Tado, responses: aioresponses) -> None:
    """Test timeout during get me."""

    # Faking a timeout by sleeping
    async def response_handler(_: str, **_kwargs: Any) -> CallbackResult:
        """Response handler for this test."""
        await asyncio.sleep(10)
        return CallbackResult(body="Goodmorning!")

    responses.get(
        f"{TADO_API_URL}/homes/1/devices",
        callback=response_handler,
    )

    with pytest.raises(TadoConnectionError):
        await python_tado.get_devices()


async def test_close_session() -> None:
    """Test not closing the session when the session does not exist."""
    tado = Tado(username="username", password="password")
    tado._close_session = True
    await tado.close()
