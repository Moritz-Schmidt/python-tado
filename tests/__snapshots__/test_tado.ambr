# serializer version: 1
# name: test_get_capabilities
  dict({
    'temperatures': dict({
      'celsius': dict({
        'max': 25.0,
        'min': 5.0,
        'step': 0.5,
      }),
      'fahrenheit': dict({
        'max': 77.0,
        'min': 41.0,
        'step': 0.5,
      }),
    }),
    'type': 'HEATING',
  })
# ---
# name: test_get_device_info
  dict({
    'celsius': 0.0,
    'fahrenheit': 0.0,
  })
# ---
# name: test_get_devices
  list([
    dict({
      'battery_state': None,
      'characteristics': dict({
        'capabilities': list([
          'RADIO_ENCRYPTION_KEY_ACCESS',
        ]),
      }),
      'child_lock_enabled': None,
      'connection_state': dict({
        'timestamp': '2024-02-27T20:13:45.407Z',
        'value': True,
      }),
      'current_fw_version': '92.1',
      'device_type': 'IB01',
      'in_pairing_mode': False,
      'mounting_state': None,
      'mounting_state_with_error': None,
      'orientation': None,
      'serial_no': 'SerialNo1',
      'short_serial_no': 'ShortSerialNo1',
    }),
    dict({
      'battery_state': 'NORMAL',
      'characteristics': dict({
        'capabilities': list([
          'INSIDE_TEMPERATURE_MEASUREMENT',
          'IDENTIFY',
        ]),
      }),
      'child_lock_enabled': True,
      'connection_state': dict({
        'timestamp': '2024-02-27T20:20:20.921Z',
        'value': True,
      }),
      'current_fw_version': '95.1',
      'device_type': 'VA02',
      'in_pairing_mode': None,
      'mounting_state': dict({
        'timestamp': '2024-02-19T17:23:30.537Z',
        'value': 'CALIBRATED',
      }),
      'mounting_state_with_error': 'CALIBRATED',
      'orientation': 'HORIZONTAL',
      'serial_no': 'SerialNo2',
      'short_serial_no': 'ShortSerialNo2',
    }),
    dict({
      'battery_state': 'NORMAL',
      'characteristics': dict({
        'capabilities': list([
          'INSIDE_TEMPERATURE_MEASUREMENT',
          'IDENTIFY',
        ]),
      }),
      'child_lock_enabled': True,
      'connection_state': dict({
        'timestamp': '2024-02-27T20:23:13.342Z',
        'value': True,
      }),
      'current_fw_version': '95.1',
      'device_type': 'VA02',
      'in_pairing_mode': None,
      'mounting_state': dict({
        'timestamp': '2024-02-26T13:33:10.647Z',
        'value': 'CALIBRATED',
      }),
      'mounting_state_with_error': 'CALIBRATED',
      'orientation': 'HORIZONTAL',
      'serial_no': 'SerialNo3',
      'short_serial_no': 'ShortSerialNo3',
    }),
    dict({
      'battery_state': 'NORMAL',
      'characteristics': dict({
        'capabilities': list([
          'INSIDE_TEMPERATURE_MEASUREMENT',
          'IDENTIFY',
        ]),
      }),
      'child_lock_enabled': True,
      'connection_state': dict({
        'timestamp': '2024-02-27T20:23:02.849Z',
        'value': True,
      }),
      'current_fw_version': '95.1',
      'device_type': 'VA02',
      'in_pairing_mode': None,
      'mounting_state': dict({
        'timestamp': '2023-10-09T15:39:08.131Z',
        'value': 'CALIBRATED',
      }),
      'mounting_state_with_error': 'CALIBRATED',
      'orientation': 'HORIZONTAL',
      'serial_no': 'SerialNo4',
      'short_serial_no': 'ShortSerialNo4',
    }),
    dict({
      'battery_state': 'NORMAL',
      'characteristics': dict({
        'capabilities': list([
          'INSIDE_TEMPERATURE_MEASUREMENT',
          'IDENTIFY',
        ]),
      }),
      'child_lock_enabled': True,
      'connection_state': dict({
        'timestamp': '2024-02-27T20:16:13.890Z',
        'value': True,
      }),
      'current_fw_version': '95.1',
      'device_type': 'VA02',
      'in_pairing_mode': None,
      'mounting_state': dict({
        'timestamp': '2024-02-04T10:17:00.266Z',
        'value': 'CALIBRATED',
      }),
      'mounting_state_with_error': 'CALIBRATED',
      'orientation': 'HORIZONTAL',
      'serial_no': 'SerialNo5',
      'short_serial_no': 'ShortSerialNo5',
    }),
  ])
# ---
# name: test_get_home_state
  dict({
    'presence': 'HOME',
    'presence_locked': True,
    'show_home_presence_switch_button': True,
    'show_switch_to_auto_geofencing_button': None,
  })
# ---
# name: test_get_me
  dict({
    'email': 'user@domain.tld',
    'homes': list([
      dict({
        'id': 1,
        'name': 'home name',
      }),
    ]),
    'id': '5',
    'locale': 'nl',
    'name': 'test_user',
    'username': 'user@domain.tld',
  })
# ---
# name: test_get_mobile_devices
  list([
    dict({
      'device_meta_data': dict({
        'locale': 'nl',
        'model': 'Model1',
        'os_version': '14',
        'platform': 'Android',
      }),
      'id': 123456,
      'location': dict({
        'at_home': True,
        'bearing_from_home': dict({
          'degrees': 21.4215315376382,
          'radians': 0.37387625615159,
        }),
        'relative_distance_from_home_fence': 0.0,
        'stale': False,
      }),
      'name': 'User1',
      'settings': dict({
        'geo_tracking_enabled': True,
        'on_demand_log_retrieval_enabled': False,
        'special_offers_enabled': True,
      }),
    }),
    dict({
      'device_meta_data': dict({
        'locale': 'nl',
        'model': 'Model2',
        'os_version': '14',
        'platform': 'Android',
      }),
      'id': 789012,
      'location': None,
      'name': 'User2',
      'settings': dict({
        'geo_tracking_enabled': False,
        'on_demand_log_retrieval_enabled': False,
        'special_offers_enabled': True,
      }),
    }),
  ])
# ---
# name: test_get_weather
  dict({
    'outside_temperature': dict({
      'celsius': 2.29,
      'fahrenheit': 36.12,
      'precision': dict({
        'celsius': 0.01,
        'fahrenheit': 0.01,
      }),
      'timestamp': '2024-02-27T20:29:49.360Z',
      'type': 'TEMPERATURE',
    }),
    'solar_intensity': dict({
      'percentage': 0.0,
      'timestamp': '2024-02-27T20:29:49.360Z',
      'type': 'PERCENTAGE',
    }),
    'weather_state': dict({
      'timestamp': '2024-02-27T20:29:49.360Z',
      'type': 'WEATHER_STATE',
      'value': 'NIGHT_CLOUDY',
    }),
  })
# ---
# name: test_get_zone_states
  list([
    dict({
      'zone_states': dict({
        '1': dict({
          'activity_data_points': dict({
            'heatingPower': dict({
              'percentage': 0.0,
              'timestamp': '2024-02-27T20:30:01.259Z',
              'type': 'PERCENTAGE',
            }),
          }),
          'geolocation_override': False,
          'geolocation_override_disable_time': None,
          'link': dict({
            'state': 'ONLINE',
          }),
          'next_schedule_change': dict({
            'setting': dict({
              'power': 'ON',
              'temperature': dict({
                'celsius': 18.0,
                'fahrenheit': 64.4,
                'precision': None,
                'timestamp': None,
                'type': None,
              }),
              'type': 'HEATING',
            }),
            'start': '2024-02-27T21:00:00Z',
          }),
          'next_time_block': dict({
            'start': '2024-02-27T21:00:00.000Z',
          }),
          'open_window': None,
          'overlay': dict({
            'projected_expiry': None,
            'setting': dict({
              'power': 'ON',
              'temperature': dict({
                'celsius': 17.0,
                'fahrenheit': 62.6,
                'precision': None,
                'timestamp': None,
                'type': None,
              }),
              'type': 'HEATING',
            }),
            'termination': dict({
              'projectedExpiry': None,
              'type': 'MANUAL',
              'typeSkillBasedApp': 'MANUAL',
            }),
            'type': 'MANUAL',
          }),
          'overlay_type': 'MANUAL',
          'preparation': None,
          'sensor_data_points': dict({
            'humidity': dict({
              'percentage': 51.0,
              'timestamp': '2024-02-27T20:37:02.553Z',
              'type': 'PERCENTAGE',
            }),
            'insideTemperature': dict({
              'celsius': 17.98,
              'fahrenheit': 64.36,
              'precision': dict({
                'celsius': 0.1,
                'fahrenheit': 0.1,
              }),
              'timestamp': '2024-02-27T20:37:02.553Z',
              'type': 'TEMPERATURE',
            }),
          }),
          'setting': dict({
            'power': 'ON',
            'temperature': dict({
              'celsius': 17.0,
              'fahrenheit': 62.6,
              'precision': None,
              'timestamp': None,
              'type': None,
            }),
            'type': 'HEATING',
          }),
          'tado_mode': 'HOME',
        }),
        '2': dict({
          'activity_data_points': dict({
            'heatingPower': dict({
              'percentage': 11.0,
              'timestamp': '2024-02-27T20:37:22.791Z',
              'type': 'PERCENTAGE',
            }),
          }),
          'geolocation_override': False,
          'geolocation_override_disable_time': None,
          'link': dict({
            'state': 'ONLINE',
          }),
          'next_schedule_change': None,
          'next_time_block': dict({
            'start': '2024-02-27T21:00:00.000Z',
          }),
          'open_window': None,
          'overlay': dict({
            'projected_expiry': None,
            'setting': dict({
              'power': 'ON',
              'temperature': dict({
                'celsius': 15.0,
                'fahrenheit': 59.0,
                'precision': None,
                'timestamp': None,
                'type': None,
              }),
              'type': 'HEATING',
            }),
            'termination': dict({
              'projectedExpiry': None,
              'type': 'MANUAL',
              'typeSkillBasedApp': 'MANUAL',
            }),
            'type': 'MANUAL',
          }),
          'overlay_type': 'MANUAL',
          'preparation': None,
          'sensor_data_points': dict({
            'humidity': dict({
              'percentage': 56.9,
              'timestamp': '2024-02-27T20:38:14.791Z',
              'type': 'PERCENTAGE',
            }),
            'insideTemperature': dict({
              'celsius': 14.89,
              'fahrenheit': 58.8,
              'precision': dict({
                'celsius': 0.1,
                'fahrenheit': 0.1,
              }),
              'timestamp': '2024-02-27T20:38:14.791Z',
              'type': 'TEMPERATURE',
            }),
          }),
          'setting': dict({
            'power': 'ON',
            'temperature': dict({
              'celsius': 15.0,
              'fahrenheit': 59.0,
              'precision': None,
              'timestamp': None,
              'type': None,
            }),
            'type': 'HEATING',
          }),
          'tado_mode': 'HOME',
        }),
        '3': dict({
          'activity_data_points': dict({
            'heatingPower': dict({
              'percentage': 0.0,
              'timestamp': '2024-02-27T20:28:21.301Z',
              'type': 'PERCENTAGE',
            }),
          }),
          'geolocation_override': False,
          'geolocation_override_disable_time': None,
          'link': dict({
            'state': 'ONLINE',
          }),
          'next_schedule_change': dict({
            'setting': dict({
              'power': 'ON',
              'temperature': dict({
                'celsius': 18.0,
                'fahrenheit': 64.4,
                'precision': None,
                'timestamp': None,
                'type': None,
              }),
              'type': 'HEATING',
            }),
            'start': '2024-02-27T21:00:00Z',
          }),
          'next_time_block': dict({
            'start': '2024-02-27T21:00:00.000Z',
          }),
          'open_window': None,
          'overlay': dict({
            'projected_expiry': None,
            'setting': dict({
              'power': 'ON',
              'temperature': dict({
                'celsius': 17.0,
                'fahrenheit': 62.6,
                'precision': None,
                'timestamp': None,
                'type': None,
              }),
              'type': 'HEATING',
            }),
            'termination': dict({
              'projectedExpiry': None,
              'type': 'MANUAL',
              'typeSkillBasedApp': 'MANUAL',
            }),
            'type': 'MANUAL',
          }),
          'overlay_type': 'MANUAL',
          'preparation': None,
          'sensor_data_points': dict({
            'humidity': dict({
              'percentage': 55.5,
              'timestamp': '2024-02-27T20:37:12.297Z',
              'type': 'PERCENTAGE',
            }),
            'insideTemperature': dict({
              'celsius': 17.52,
              'fahrenheit': 63.54,
              'precision': dict({
                'celsius': 0.1,
                'fahrenheit': 0.1,
              }),
              'timestamp': '2024-02-27T20:37:12.297Z',
              'type': 'TEMPERATURE',
            }),
          }),
          'setting': dict({
            'power': 'ON',
            'temperature': dict({
              'celsius': 17.0,
              'fahrenheit': 62.6,
              'precision': None,
              'timestamp': None,
              'type': None,
            }),
            'type': 'HEATING',
          }),
          'tado_mode': 'HOME',
        }),
        '4': dict({
          'activity_data_points': dict({
            'heatingPower': dict({
              'percentage': 0.0,
              'timestamp': '2024-02-27T20:27:18.903Z',
              'type': 'PERCENTAGE',
            }),
          }),
          'geolocation_override': False,
          'geolocation_override_disable_time': None,
          'link': dict({
            'state': 'ONLINE',
          }),
          'next_schedule_change': dict({
            'setting': dict({
              'power': 'ON',
              'temperature': dict({
                'celsius': 18.0,
                'fahrenheit': 64.4,
                'precision': None,
                'timestamp': None,
                'type': None,
              }),
              'type': 'HEATING',
            }),
            'start': '2024-02-27T21:00:00Z',
          }),
          'next_time_block': dict({
            'start': '2024-02-27T21:00:00.000Z',
          }),
          'open_window': None,
          'overlay': dict({
            'projected_expiry': None,
            'setting': dict({
              'power': 'OFF',
              'temperature': None,
              'type': 'HEATING',
            }),
            'termination': dict({
              'projectedExpiry': None,
              'type': 'MANUAL',
              'typeSkillBasedApp': 'MANUAL',
            }),
            'type': 'MANUAL',
          }),
          'overlay_type': 'MANUAL',
          'preparation': None,
          'sensor_data_points': dict({
            'humidity': dict({
              'percentage': 63.3,
              'timestamp': '2024-02-27T20:37:20.931Z',
              'type': 'PERCENTAGE',
            }),
            'insideTemperature': dict({
              'celsius': 18.57,
              'fahrenheit': 65.43,
              'precision': dict({
                'celsius': 0.1,
                'fahrenheit': 0.1,
              }),
              'timestamp': '2024-02-27T20:37:20.931Z',
              'type': 'TEMPERATURE',
            }),
          }),
          'setting': dict({
            'power': 'OFF',
            'temperature': None,
            'type': 'HEATING',
          }),
          'tado_mode': 'HOME',
        }),
      }),
    }),
  ])
# ---
# name: test_get_zones
  list([
    dict({
      'date_created': '2023-04-12T12:58:12.737Z',
      'dazzle_enabled': True,
      'dazzle_mode': dict({
        'enabled': True,
        'supported': True,
      }),
      'device_types': list([
        'VA02',
      ]),
      'devices': list([
        dict({
          'battery_state': 'NORMAL',
          'characteristics': dict({
            'capabilities': list([
              'INSIDE_TEMPERATURE_MEASUREMENT',
              'IDENTIFY',
            ]),
          }),
          'child_lock_enabled': True,
          'connection_state': dict({
            'timestamp': '2024-02-27T20:30:13.976Z',
            'value': True,
          }),
          'current_fw_version': '95.1',
          'device_type': 'VA02',
          'in_pairing_mode': None,
          'mounting_state': dict({
            'timestamp': '2024-02-04T10:17:00.266Z',
            'value': 'CALIBRATED',
          }),
          'mounting_state_with_error': 'CALIBRATED',
          'orientation': 'HORIZONTAL',
          'serial_no': 'Serial1',
          'short_serial_no': 'ShortSerial1',
        }),
      ]),
      'id': 2,
      'name': 'Zone1',
      'open_window_detection': dict({
        'enabled': False,
        'supported': True,
        'timeout_in_seconds': 900,
      }),
      'report_available': False,
      'show_schedule_detup': False,
      'supports_dazzle': True,
      'type': 'HEATING',
    }),
    dict({
      'date_created': '2023-01-29T16:02:14.530Z',
      'dazzle_enabled': True,
      'dazzle_mode': dict({
        'enabled': True,
        'supported': True,
      }),
      'device_types': list([
        'VA02',
      ]),
      'devices': list([
        dict({
          'battery_state': 'NORMAL',
          'characteristics': dict({
            'capabilities': list([
              'INSIDE_TEMPERATURE_MEASUREMENT',
              'IDENTIFY',
            ]),
          }),
          'child_lock_enabled': True,
          'connection_state': dict({
            'timestamp': '2024-02-27T20:32:05.188Z',
            'value': True,
          }),
          'current_fw_version': '95.1',
          'device_type': 'VA02',
          'in_pairing_mode': None,
          'mounting_state': dict({
            'timestamp': '2023-10-09T15:39:08.131Z',
            'value': 'CALIBRATED',
          }),
          'mounting_state_with_error': 'CALIBRATED',
          'orientation': 'HORIZONTAL',
          'serial_no': 'Serial2',
          'short_serial_no': 'ShortSerial2',
        }),
      ]),
      'id': 1,
      'name': 'Zone2',
      'open_window_detection': dict({
        'enabled': False,
        'supported': True,
        'timeout_in_seconds': 900,
      }),
      'report_available': False,
      'show_schedule_detup': True,
      'supports_dazzle': True,
      'type': 'HEATING',
    }),
    dict({
      'date_created': '2023-04-14T07:52:56.352Z',
      'dazzle_enabled': True,
      'dazzle_mode': dict({
        'enabled': True,
        'supported': True,
      }),
      'device_types': list([
        'VA02',
      ]),
      'devices': list([
        dict({
          'battery_state': 'NORMAL',
          'characteristics': dict({
            'capabilities': list([
              'INSIDE_TEMPERATURE_MEASUREMENT',
              'IDENTIFY',
            ]),
          }),
          'child_lock_enabled': True,
          'connection_state': dict({
            'timestamp': '2024-02-27T20:31:11.417Z',
            'value': True,
          }),
          'current_fw_version': '95.1',
          'device_type': 'VA02',
          'in_pairing_mode': None,
          'mounting_state': dict({
            'timestamp': '2024-02-26T13:33:10.647Z',
            'value': 'CALIBRATED',
          }),
          'mounting_state_with_error': 'CALIBRATED',
          'orientation': 'HORIZONTAL',
          'serial_no': 'Serial3',
          'short_serial_no': 'ShortSerial3',
        }),
      ]),
      'id': 3,
      'name': 'Zone3',
      'open_window_detection': dict({
        'enabled': False,
        'supported': True,
        'timeout_in_seconds': 900,
      }),
      'report_available': False,
      'show_schedule_detup': True,
      'supports_dazzle': True,
      'type': 'HEATING',
    }),
    dict({
      'date_created': '2023-04-14T07:58:45.196Z',
      'dazzle_enabled': True,
      'dazzle_mode': dict({
        'enabled': True,
        'supported': True,
      }),
      'device_types': list([
        'VA02',
      ]),
      'devices': list([
        dict({
          'battery_state': 'NORMAL',
          'characteristics': dict({
            'capabilities': list([
              'INSIDE_TEMPERATURE_MEASUREMENT',
              'IDENTIFY',
            ]),
          }),
          'child_lock_enabled': True,
          'connection_state': dict({
            'timestamp': '2024-02-27T20:33:21.903Z',
            'value': True,
          }),
          'current_fw_version': '95.1',
          'device_type': 'VA02',
          'in_pairing_mode': None,
          'mounting_state': dict({
            'timestamp': '2024-02-19T17:23:30.537Z',
            'value': 'CALIBRATED',
          }),
          'mounting_state_with_error': 'CALIBRATED',
          'orientation': 'HORIZONTAL',
          'serial_no': 'Serial4',
          'short_serial_no': 'ShortSerial4',
        }),
      ]),
      'id': 4,
      'name': 'Zone4',
      'open_window_detection': dict({
        'enabled': False,
        'supported': True,
        'timeout_in_seconds': 900,
      }),
      'report_available': False,
      'show_schedule_detup': True,
      'supports_dazzle': True,
      'type': 'HEATING',
    }),
  ])
# ---
